com.unity.ml-agents.extensions/Runtime/Sensors/GridSensor.cs

```diff
@@ -536,73 +551,119 @@ protected void ChannelsToTexture(int channelIndex, int numChannelsToAdd)
+        /// <returns></returns>
+        int BufferResizingOverlapBoxNonAlloc(Vector3 cellCenter, Vector3 halfCellScale, Quaternion rotation)
+        {
+            int numFound;
+            while (true)
+            {
+                numFound = Physics.OverlapBoxNonAlloc(cellCenter, halfCellScale, m_CollidersPerCell, rotation, ObserveMask);
+                if (numFound == m_CollidersPerCell.Length && m_CollidersPerCell.Length < k_AbsoluteMaxCollidersPerCell)
+                {
+                    m_CollidersPerCell = new Collider[Math.Min(k_AbsoluteMaxCollidersPerCell, m_CollidersPerCell.Length * 2)];
+                    EstimatedMaxCollidersPerCell = m_CollidersPerCell.Length;
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            return numFound;
+        }
+
         /// <summary>
         /// Parses the array of colliders found within a cell. Finds the closest gameobject to the agent root reference within the cell
         /// </summary>
         /// <param name="foundColliders">Array of the colliders found within the cell</param>
+        /// <param name="numFound">Number of colliders found.</param>
         /// <param name="cellIndex">The index of the cell</param>
         /// <param name="cellCenter">The center position of the cell</param>
-        protected virtual void ParseColliders(Collider[] foundColliders, int cellIndex, Vector3 cellCenter)
+        protected virtual void ParseColliders(Collider[] foundColliders, int numFound, int cellIndex, Vector3 cellCenter)
         {
-            GameObject currentColliderGo = null;
+            Profiler.BeginSample("GridSensor.ParseColliders");
             GameObject closestColliderGo = null;
-            Vector3 closestColliderPoint = Vector3.zero;
-            float distance = float.MaxValue;
-            float currentDistance = 0f;
+            var distance = float.MaxValue;
 
-            for (int i = 0; i < foundColliders.Length; i++)
+            for (var i = 0; i < numFound; i++)
             {
-                currentColliderGo = foundColliders[i].gameObject;
+                var currentColliderGo = foundColliders[i].gameObject;
 
                 // Continue if the current collider go is the root reference
-                if (currentColliderGo == rootReference)
+                if (ReferenceEquals(currentColliderGo, rootReference))
                     continue;
 
-                closestColliderPoint = foundColliders[i].ClosestPointOnBounds(cellCenter);
-                currentDistance = Vector3.Distance(closestColliderPoint, rootReference.transform.position);
+                Profiler.BeginSample("ClosestPointOnBounds");
+                var closestColliderPoint = foundColliders[i].ClosestPointOnBounds(cellCenter);
+                Profiler.EndSample();
+                var currentDistance = Vector3.Distance(closestColliderPoint, rootReference.transform.position);
```

Comment:
Small optimization - you can used distance squared here
`(closestColliderPoint - rootReference.transform.position).SqrMagnitude()`
and take a sqrt() before calling LoadObjectData instead.